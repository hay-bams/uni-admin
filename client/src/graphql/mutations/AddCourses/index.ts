import { gql } from "@apollo/client";

export const ADD_COURSES = gql`
  mutation AddCourse($studentId: String!, $input: ID!) {
    addCourses(studentId: $studentId, input: $input) {
      id
    }
  }
`;