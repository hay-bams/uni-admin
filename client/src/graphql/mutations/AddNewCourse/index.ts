import { gql } from "@apollo/client";

export  const ADD_NEW_COURSE = gql`
  mutation AddNewCourse($input: NewCourseInput) {
    addNewCourse(input: $input) {
      id
    }
  }
`