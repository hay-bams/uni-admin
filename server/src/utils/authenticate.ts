import { ObjectID } from 'mongodb';
import jwt from 'jsonwebtoken'
import {  ICtx } from '../lib/types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const Authenticated =  <T=any, K=any> (next: any) => async (
  root: T,
  args: K,
  ctx: ICtx
) => {
  const { db, req, res } = ctx;
  const token = req.get("X-CSRF-TOKEN");

  const id_cookie = req.signedCookies.admin;
  const decoded = jwt.verify(`${token}`, `${process.env.SECRET}`) as { admin: string }

  if(id_cookie !== decoded.admin) {
    throw new Error('This may be a CSRF attack, please sign in')
  }

  const admin = await db.users.findOne({
    _id: new ObjectID(id_cookie),
  });

  if (!admin) {
    res.clearCookie('admin');
    throw new Error('User not authenticated, please sign in first');
  }

  return next(root, args, ctx);
};
