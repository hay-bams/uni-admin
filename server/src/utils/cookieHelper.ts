import { Request, Response } from "express"
import { ObjectID } from "mongodb"

import { Database } from "../lib/types"

export const cookieOptions = {
  httpOnly: true,
  sameSite: true,
  signed: true,
  secure: process.env.NODE_ENV === 'development' ? false : true
}

export const cookie = {
  setCookie: (id: ObjectID | undefined, res: Response) => {
    if(id) {
      res.cookie('admin', id, {
        ...cookieOptions,
        maxAge: 1000 * 60 * 60 * 24 * 365
      })
    }
  },
  loginViaCookie: async (db: Database, req: Request, res: Response) => {
    const result = await db.users.findOne({
      _id: new ObjectID(req.signedCookies.admin),
    });
  
    if (!result) {
      res.clearCookie('admin');
      return null;
    }
  
    return result;
  }
}