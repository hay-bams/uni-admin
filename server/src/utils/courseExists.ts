import { ObjectID } from "mongodb";
import { Course, Database, Student } from "../lib/types";

export const curseExists = async (
  ids: string[],
  student: Student,
  db: Database
): Promise<Course | null> => {
  const courses = student.courses;
  const map = new Map();

  for (const course of courses) {
    map.set(course.toString(), course.toString());
  }

  for (const id of ids) {
    if (map.get(id)) {
      const course = await db.courses.findOne({
        _id: new ObjectID(map.get(id))
      })

      if(course) return course;
    }
  }
  return null;
};
