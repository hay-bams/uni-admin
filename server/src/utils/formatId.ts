import { ObjectID } from "mongodb"

export const formatId = (ids: string[]) => {
  const newIds = ids.map((id) => {
    return new ObjectID(id)
  })

  return newIds
}