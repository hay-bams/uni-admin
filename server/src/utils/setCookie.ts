import { Response } from "express"
import { ObjectID } from "mongodb"

export const cookieOptions = {
  httpOnly: true,
  sameSite: true,
  signed: true,
  secure: process.env.NODE_ENV === 'development' ? false : true
}

export const setCookie = (id: ObjectID | undefined, res: Response) => {
  if(id) {
    res.cookie('user', id, {
      ...cookieOptions,
      maxAge: 1000 * 60 * 60 * 24 * 365
    })
  }
}