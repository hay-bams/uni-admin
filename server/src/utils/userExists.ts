import { Database } from "../lib/types";

export const userExist = async (db: Database, username: string) => {
  const user = await db.users.findOne({
    username,
  });

  if (user) {
    return true;
  }

  return false;
};