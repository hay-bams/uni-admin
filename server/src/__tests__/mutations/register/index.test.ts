/* eslint-disable @typescript-eslint/no-non-null-assertion */
import bcrypt from 'bcryptjs';

import { USER_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import { TEST_REGISTER } from '../mutations';
import { cookie } from '../../../utils/cookieHelper';

jest.mock('bcryptjs');
const mockedBcrypt = bcrypt as jest.Mocked<typeof bcrypt>;
const bcryptCompare = mockedBcrypt.compare;

beforeEach(() => {
  bcryptCompare.mockReset();
});

describe('Admin Register Mutation', () => {
  test('should register admin', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          insertOne: jest.fn(() => ({ ops: [USER_DATA[0]] })),
          findOne: jest.fn(() => null),
        },
      },
    });

    jest.spyOn(bcrypt, 'compare').mockImplementation(jest.fn(() => true));
    jest.spyOn(cookie, 'setCookie').mockImplementation(jest.fn());

    const res = await mutate({
      mutation: TEST_REGISTER,
      variables: {
        input: {
          username: 'haybams',
          password: 'password',
        },
      },
    });

    expect(res.data.register).toHaveProperty('id');
    expect(res.data.register).toHaveProperty('username');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if user already exists', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
    });

    jest.spyOn(bcrypt, 'compare').mockImplementation(jest.fn(() => true));
    jest.spyOn(cookie, 'setCookie').mockImplementation(jest.fn());

    const res = await mutate({
      mutation: TEST_REGISTER,
      variables: {
        input: {
          username: 'haybams',
          password: 'password',
        },
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });
});
