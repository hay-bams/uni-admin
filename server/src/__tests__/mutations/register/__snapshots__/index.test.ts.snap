// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Register Mutation should register admin 1`] = `
Object {
  "data": Object {
    "register": Object {
      "id": "5fe09161251c8b0a9565108d",
      "username": "admin",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Admin Register Mutation should throw an error if user already exists 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: user already exists],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
